core/trino-main/src/test/java/io/trino/execution/scheduler/TestScaledWriterScheduler.java

```diff
@@ -0,0 +1,421 @@
+                        "some_node",
+                        ImmutableList.of(),
+                        0,
+                        0,
+                        new OutputBufferStatus(OptionalLong.empty(), isOutputBufferOverUtilized, false),
+                        DataSize.of(32, DataSize.Unit.MEGABYTE),
+                        DataSize.of(1, DataSize.Unit.MEGABYTE),
+                        DataSize.of(1, DataSize.Unit.MEGABYTE),
+                        DataSize.of(0, DataSize.Unit.MEGABYTE),
+                        0,
+                        Duration.valueOf("0s"),
+                        0,
+                        1,
+                        1),
+                DateTime.now(),
+                new OutputBufferInfo(
+                        "TESTING",
+                        BufferState.FINISHED,
+                        false,
+                        false,
+                        0,
+                        0,
+                        0,
+                        0,
+                        Optional.empty(),
+                        Optional.of(new TDigestHistogram(new TDigest())),
+                        Optional.empty()),
+                ImmutableSet.of(),
+                new TaskStats(DateTime.now(),
+                        null,
+                        null,
+                        null,
+                        null,
+                        new Duration(0, MILLISECONDS),
+                        new Duration(0, MILLISECONDS),
+                        0,
+                        0,
+                        0,
+                        0L,
+                        0,
+                        0,
+                        0L,
+                        0,
+                        0,
+                        0.0,
+                        DataSize.ofBytes(0),
+                        DataSize.ofBytes(0),
+                        DataSize.ofBytes(0),
+                        new Duration(0, MILLISECONDS),
+                        new Duration(0, MILLISECONDS),
+                        new Duration(0, MILLISECONDS),
+                        false,
+                        ImmutableSet.of(),
+                        DataSize.ofBytes(0),
+                        0,
+                        new Duration(0, MILLISECONDS),
+                        DataSize.ofBytes(0),
+                        0,
+                        DataSize.ofBytes(0),
+                        0,
+                        DataSize.ofBytes(0),
+                        0,
+                        new Duration(0, MILLISECONDS),
+                        DataSize.ofBytes(outputDataSize),
+                        0,
+                        new Duration(0, MILLISECONDS),
+                        DataSize.ofBytes(0),
+                        0,
+                        new Duration(0, MILLISECONDS),
+                        ImmutableList.of()),
+                Optional.empty(),
+                false);
+    }
+
+    private static class TestingStageExecution
```

Comment:
This is very similar to the class there in `TestPhasedExecutionSchedule`

Maybe it's better to extract the `TestingStageExecution` class out of `TestPhasedExecutionSchedule` and use it in both places? WDYT?

Reply:
based on your previous comment, you wanted not to have association between unit tests, so recommended copying it even though there will be a lot of code duplication. I would like to stick to it.

Reply:
In that case, do we have to implement methods like `getFragment()` etc

Reply:
how can we skip implementing them since it is inheriting an interface?

Reply:
I think we can simply return unsupported exceptions since those methods are not used in `ScaleWriterScheduler`

maybe something like this:

```
        public TestingStageExecution()
        {}

        @Override
        public PlanFragment getFragment()
        {
            throw new UnsupportedOperationException();
        }

        @Override
        public boolean isAnyTaskBlocked()
        {
            throw new UnsupportedOperationException();
        }
        ...

        @Override
        public Optional<RemoteTask> scheduleTask(InternalNode node, int partition, Multimap<PlanNodeId, Split> initialSplits)
        {
            return Optional.of(new TestingRemoteTask(TaskId.valueOf("taskId"), "nodeId", dummyFragment()));
        }
```

Reply:
It still needs PlanFragment. And does it really matter since it's a mock? I feel like there are too many comments around unit test itself that are not directly related to the actual improvement(e.g. naming preferences) and are boilerplate codes due to not using Mockito(e.g. copying Mock implementation here and there and changing them, removing them again, etc...) that seem making the test super delicate to future changes.

Reply:
> I feel like there are too many discussions around unit test itself that are not related to the actual improvement(e.g. naming preferences)

Overall, the implementation looks good to me. @sopel39 can you also take a look?

> are boilerplate codes due to not using Mockito(e.g. copying Mock implementation here and there and changing them, removing them again, etc...) that seem making the test super delicate to future changes.

It's part of Trino's guidelines that we avoid mocking libraries.  Not sure if you have seen this.
https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#avoid-mocks

Reply:
I have seen it and I will try to follow it regardless of whether I agree or not. Just too much effort on testing mock implementations would be superfluous.